finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,719 @@
+        let handler = &mut handlers[0];
+        let messages = discovery.discover_authorities(handler);
+        assert!(messages.is_empty());
+    }
+
+    #[tokio::test]
+    async fn rebroadcasts_and_accepts_addresses() {
+        let (mut discovery, mut handlers) = build().await;
+        let authentication = handlers[1].authentication();
+        let handler = &mut handlers[0];
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        assert_eq!(addresses.len(), authentication.0.addresses().len());
+        assert_eq!(
+            addresses[0].encode(),
+            authentication.0.addresses()[0].encode()
+        );
+        match maybe_command {
+            Some((
+                DiscoveryMessage::AuthenticationBroadcast(rebroadcast_authentication),
+                DataCommand::Broadcast,
+            )) => assert_eq!(authentication.encode(), rebroadcast_authentication.encode()),
+            _ => panic!("Expected a rebroadcast, got {:?}", maybe_command),
+        }
+    }
+
+    #[tokio::test]
+    async fn does_not_rebroadcast_wrong_authentications() {
+        let (mut discovery, mut handlers) = build().await;
+        let (auth_data, _) = handlers[1].authentication();
+        let (_, signature) = handlers[2].authentication();
+        let authentication = (auth_data, signature);
+        let handler = &mut handlers[0];
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication),
+            handler,
+        );
+        assert!(addresses.is_empty());
+        assert!(maybe_command.is_none());
+    }
+
+    #[tokio::test]
+    async fn does_not_rebroadcast_quickly() {
+        let (mut discovery, mut handlers) = build().await;
+        let authentication = handlers[1].authentication();
+        let handler = &mut handlers[0];
+        discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        assert_eq!(addresses.len(), authentication.0.addresses().len());
+        assert_eq!(
+            addresses[0].encode(),
+            authentication.0.addresses()[0].encode()
+        );
+        assert!(maybe_command.is_none());
+    }
+
+    #[tokio::test]
+    async fn rebroadcasts_after_cooldown() {
+        let (mut discovery, mut handlers) = build().await;
+        let authentication = handlers[1].authentication();
+        let handler = &mut handlers[0];
+        discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        sleep(Duration::from_millis(MS_COOLDOWN + 5));
```

Comment:
I don't really like the idea of using sleep in tests, but I'm not sure if we can overcome this with lost cost

Reply:
+1 for taste, but the only reasonable solution for this I see would be to expose some (private) methods within `Discovery` on which we can listen asynchronously. But probably in this case this is an overkill

Reply:
I really don't have better ideas here, as this feature is entirely time-based. :<

Reply:
Theoretically you can mock time suppliers, but it is probably an overkill

Reply:
I agree this would be overkill, resolving as per private communication.